### DISCUSSION TITLE
[lit/ssr]: emit an error event on lit hydration errors

### DISCUSSION THREAD

**Discussion by KonnorRogers** [lit/lit#4703] [Ideas]
I was curious instead of just `throw Error()` in Lit SSR, if it'd be possible to emit an error prior to throwing.

Something like:

```js
const LitHydrationError = new Event("lit-hydration-error", { bubbles: true, composed: true, cancelable: false })

LitHydrationError.message = "Hydration Mismatch"

element.dispatchEvent(LitHydrationError)
throw new Error(LitHydrationError.message)
```

My use-case is being able to diff between client-only components and server-rendered / client-hydrated components and seeing where the mismatches are happening.

Right now I have a small POC with a `try {} catch {}` in the `updated` function thats a broad catch for errors which may not necessarily have anything to do with hydration and could be a strict rendering error.

**Comment by augustjk** [comment]
That's an interesting idea. Any improvement that helps the debug process would be great.

We have also considered a mode of hydration where mismatch would not result in an error thrown but clearing out the server-rendered DOM and re-rendering on the client. https://github.com/lit/lit/issues/1434#issuecomment-1601410576
